tasks/libs/ciproviders/gitlab_api.py

```diff
@@ -513,6 +517,60 @@ def filter_yaml(key, value):
     return {node[0]: node[1] for node in (filter_yaml(k, v) for k, v in yml.items()) if node is not None}
 
 
+def expand_matrix_jobs(yml: dict) -> dict:
+    """
+    Will expand matrix jobs into multiple jobs
+    """
+    new_jobs = {}
+    to_remove = set()
+    for job in yml:
+        if 'parallel' in yml[job] and 'matrix' in yml[job]['parallel']:
+            to_remove.add(job)
+            for arg in yml[job]['parallel']['matrix']:
```

Comment:
- I would suggest this to be in a dedicated method with some tests
- I think your implementation is complex (but I maybe misunderstood something). Suggestion:
```
value_lists = []
for value in arg.values():
    if isinstance(value, str):
        value_lists.append([value])
    else:
        value_lists.append(value)
all_args = list(prod(*value_lists))
```
And to set variables to jobs you can zip the `arg.keys()` with each element of `all_args` (they are normally in the correct order)
wdyt?